Generalize split method to work with elements of any size

austinltam opened the pull request:
ref #15115

bwspenc commented:
##########################################################################
ERROR: MOOSE prefers two spaces instead of tabs. The following files contain tab characters:
	modules/xfem/src/efa/EFAFragment2D.C
##########################################################################

bwspenc commented:
I think running clang-format will take care of those tabs:
```
git clang-format HEAD~1

bwspenc commented on modules/xfem/src/efa/EFAFragment2D.C:
```diff
@@ -334,355 +334,69 @@ EFAFragment2D::split()
-    unsigned int iedge = 0;
-    unsigned int icutedge = 0;
+      if (frag_number == 0)
+      {
```
Our convention is to not use curly braces around one-line blocks. You have a bunch of these -- just delete the opening and closing braces.

bwspenc commented:
@jiangwen84 This may interest you.

jiangwen84 commented:
@bwspenc no new test?

bwspenc commented:
@jiangwen84 I don't think we need new tests on this. This is mostly just a cleanup in preparation for other work that will add new capabilities. This stuff is tested quite a bit already.

jiangwen84 commented on modules/xfem/src/efa/EFAFragment2D.C:
```diff
@@ -334,355 +334,69 @@ EFAFragment2D::split()
-  {
-    EFAError("In split() fragment cannot have more than 2 cut edges");
-  }
-  else if (cut_edges.size() == 1 && cut_nodes.size() == 1)
```
This part of code was to deal with the cutting through nodes cases. Do we still have that capability?

bwspenc replied:
He says that he handles that. I need to look through the code carefully and convince myself about that. The tests supposedly pass, which is a good sign.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/507392/) on 0fe3447 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15118/site)

This comment will be updated on new commits.

bwspenc commented:
A lot of those failed tests are unrelated to your commit, and (I assume) someone else is probably resolving the underlying issues and we'll just need to resubmit once those are fixed. However, some of these are actual failures related to your code:
https://civet.inl.gov/job/507388/

bwspenc commented:
Looks like this is causing a segfault on some of these platforms. For example: 
```
[----------] 31 tests from ElementFragmentAlgorithm
[ RUN      ] ElementFragmentAlgorithm.test1a
./run_tests: line 26: 16025 Segmentation fault      ./$APPLICATION_NAME-unit-$METHOD $*

bwspenc commented:
chmod -x filename

bwspenc commented on modules/xfem/src/efa/EFAFragment2D.C:
```diff
@@ -334,354 +334,83 @@ EFAFragment2D::split()
-    }
-    new_fragments.push_back(new_frag);
+    EFAFragment2D * new_frag0 = new EFAFragment2D(_host_elem, false, NULL);
+    fragment_nodes0.insert(fragment_nodes0.end(), fragment_nodes1.begin + 1, fragment_nodes1.end());
```
It should be `fragment_nodes1.begin()`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/527987/) on 610356e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15118/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15118/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 77b75378dd3cd3e4eff8741383a1ff9f847c3ded`

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 30d38fc9b5bc608655d033117a8d5cb59c57ae4c
+Subproject commit 179e8c8c2dcbfcc01311bd6f5044a9534aa21584
```
**Caution!** This contains a submodule update

bwspenc commented:
@austinltam You need to get rid of the libmesh submodule update in this PR (which is a bit of a pain to do -- let me know if you need help with that).

loganharbour commented:
Trying to get our open PR count down.

Updated the submodule change to current to avoid a force push.

austinltam committed:
EFAFragment2D update ref #15115

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/553779/) on d3f8293 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15118/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15118/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 615a1458450540f0b90fff1fdf45f73d1ad08bac`

bwspenc committed:
Fix const issue to allow for debugging output ref #15115

bwspenc committed:
Update gold files for tests ref #15115

These results all changed because of differences in the ordering of the
newly-created elements. I verified that the solutions are unchanged.

I also slightly loosened a test tolerance.

bwspenc commented:
@jiangwen84 FYI, I think this is sorted out now.  I spent a while with @austinltam sorting out the test failures today. I think the code is ready to go. This does change the ordering of the creation of new elements in some cases, so I had to update some of the tests.

dschwen commented on modules/xfem/src/efa/EFAFragment2D.C:
```diff
@@ -334,345 +334,86 @@ EFAFragment2D::split()
-      new_fragments.push_back(new_frag1);
-      new_fragments.push_back(new_frag2);
+      ++node_cut_count;
+      if (frag_number == 0)
+        frag_number = 1;
+      else
+        frag_number = 0;
```
these could be replaced by 

```
frag_number = 1 - frag_number;
```

bwspenc replied:
That wouldn't work if `frag_number` is 0, would it?

dschwen commented on modules/xfem/src/efa/EFAFragment2D.C:
```diff
@@ -334,345 +334,86 @@ EFAFragment2D::split()
+  if ((edge_cut_count + node_cut_count) > 1) // any two cuts case
+  {
+    EFAFragment2D * new_frag0 = new EFAFragment2D(_host_elem, false, NULL);
+    EFAFragment2D * new_frag1 = new EFAFragment2D(_host_elem, false, NULL);
```
I'd love to see smart pointer here. What is the life time management of these objects?

bwspenc replied:
I agree. I made an issue to switch everything in this code over to smart pointers about 5 years ago, and still need to get to that...

These are indeed managed correctly (deleted by the elements associated with them), and the code runs cleanly in valgrind, but smart pointers would be nice. That's way beyond the scope of this changeset, though.

dschwen commented on modules/xfem/src/efa/EFAFragment2D.C:
```diff
@@ -334,345 +334,86 @@ EFAFragment2D::split()
+    if (fragment_nodes[1].size() >= 3)
+    { // check to make sure the cut wasn't straight on an edge
+      for (unsigned int inode = 0; inode < fragment_nodes[1].size() - 1;
+           inode++) // assemble fragment 2 (1)
+        new_frag1->addEdge(new EFAEdge(fragment_nodes[1][inode], fragment_nodes[1][inode + 1]));
 
-      new_frag2->addEdge(new EFAEdge(node2, node3));
-      new_frag2->addEdge(new EFAEdge(node3, node4));
-      new_frag2->addEdge(new EFAEdge(node4, node2));
+      new_frag1->addEdge(
+          new EFAEdge(fragment_nodes[1][fragment_nodes[1].size() - 1], fragment_nodes[1][0]));
```
This block is the same code as the block above (just with 1 instead of 0) This could be made more concise (and maintainable with a two step for loop)

bwspenc replied:
Good point! I just pushed an update that cleans that up.

bwspenc committed:
Address review comments ref #15115

bwspenc commented:
@dschwen The update I just pushed addresses the code duplication. The smart pointer conversion is a much larger issue, and I'm not sure what you mean about using `frag_number-1`. If `frag_number` is zero, that just makes it have the maximum value of an unsigned int. If it was a data type that could just be 0 or 1, that would work. I agree that it would be nice to do that in one line, so if I'm missing something on how to do that, please let me know.

dschwen commented:
You got the frag number thing backwards

1 - 0 -> 1 
1 - 1 -> 0

That line would flip between 1 and 0

bwspenc commented:
Thanks, @dschwen!  Not sure how I missed that. That's a pretty cool way to do the toggling! I made the change.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/554267/) on 21b4bf6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15118/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15118/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 615a1458450540f0b90fff1fdf45f73d1ad08bac`

jiangwen84 reviewed (dismissed):
Looks good.

bwspenc commented on modules/xfem/src/efa/EFAFragment2D.C:
```diff
@@ -334,345 +334,70 @@ EFAFragment2D::split()
-  std::vector<unsigned int> cut_edges;
-  std::vector<unsigned int> cut_nodes;
+  std::vector<std::vector<EFANode *>> fragment_nodes; // vectors of EFA nodes in the two fragments
+  fragment_nodes.resize(2);
```
Yes. Just pushed an update that makes that change.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/554467/) on ee8f460 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15118/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15118/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 615a1458450540f0b90fff1fdf45f73d1ad08bac`

loganharbour reviewed (approved):
👍

lindsayad commented:
This is failing XFEM valgrind testing. See for instance: https://civet.inl.gov/event/56402/

bwspenc commented:
Uggh!  I'll look into it.
